README.md
# Sensirion Raspberry Pi I2C SDP Driver

This document explains how to set up the SDP sensor to run on a Raspberry Pi
using the provided code.

<center><img src="images/SDP8xx.png" width="250px"><img src="images/SDP3x.png" width="250px"></center>


Click [here](https://www.sensirion.com/en/flow-sensors/differential-pressure-sensors/test-your-sdp3x-differential-pressure-sensor-with-the-evaluation-kit-ek-p4/) to learn more about the SDP3x Sensor Evaluation Kit.

Click [here](https://www.sensirion.com/en/flow-sensors/differential-pressure-sensors/evaluation-kit-for-sensors-of-the-sdp800-series/) to learn more about the SDP800 series Sensor Evaluation Kit.


## Setup Guide

### Connecting the Sensor

Your sensor has the four different connectors: VCC, GND, SDA, SCL. Use
the following pins to connect your Sdp:

 *Sdp*  |    *Raspberry Pi*
 :------: | :------------------:
   VCC    |        Pin 1
   GND    |        Pin 6
   SDA    |        Pin 3
   SCL    |        Pin 5

<center><img src="images/GPIO-Pinout-Diagram.png" width="900px"></center>

### Setting sensor address ###

Depending on whether you connect a SDP3x or SDP8xx configure the correct i2c address in sdp_i2c.h (SDP_I2C_ADDRESS).

### Raspberry Pi

- [Install the Raspberry Pi OS on to your Raspberry Pi](https://projects.raspberrypi.org/en/projects/raspberry-pi-setting-up)
- [Enable the I2C interface in the raspi-config](https://www.raspberrypi.org/documentation/configuration/raspi-config.md)
- Download the driver for the [Sensirion Github Page](https://github.com/Sensirion/raspberry-pi-i2c-sdp) and extract the `.zip` on your Raspberry Pi
- Compile the driver
    1. Open a [terminal](https://www.raspberrypi.org/documentation/usage/terminal/?)
    2. Navigate to the driver directory. E.g. `cd ~/raspberry-pi-i2c-sdp`
    3. Run the `make` command to compile the driver

       Output:
       ```
       rm -f sdp_i2c_example_usage
       cc -Os -Wall -fstrict-aliasing -Wstrict-aliasing=1 -Wsign-conversion -fPIC -I. -o sdp_i2c_example_usage  sdp_i2c.h sdp_i2c.c sensirion_i2c_hal.h sensirion_i2c.h sensirion_i2c.c \
           sensirion_i2c_hal.c sensirion_config.h sensirion_common.h sensirion_common.c sdp_i2c_example_usage.c
       ```
- Test your connected sensor
    - Run `./sdp_i2c_example_usage` in the same directory you used to
      compile the driver.

      Output:
      ```
      Product number: 0x03020b01
      Serial Number: 1902067650
      Differential pressure: -0.17 Pa
      Temperature: 28.80 °C
      Differential pressure: -0.03 Pa
      Temperature: 28.80 °C
      Differential pressure: -0.03 Pa
      Temperature: 28.80 °C
      Differential pressure: -0.03 Pa
      ...
      ```

## Troubleshooting

### Building driver failed

If the execution of `make` in the compilation step 3 fails with something like

> -bash: make: command not found

your RaspberryPi likely does not have the build tools installed. Proceed as follows:

```
$ sudo apt-get update
$ sudo apt-get upgrade
$ sudo apt-get install build-essential
```

### Initialization failed

```
Error executing sdp_get_serial_number(): -1
Error executing sdp_read_product_identifier(): -1
Error executing sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging(): -1
Error executing sdp_read_measurement(): -1
```

then go through the below troubleshooting steps.

-   Ensure that you connected the sensor correctly: All cables are fully
    plugged in and connected to the correct pin.
-   Ensure that I2C is enabled on the Raspberry Pi. For this redo the steps on
    "Enable the I2C interface in the raspi-config" in the guide above.
-   Ensure that your user account has read and write access to the I2C device.
    If it only works with user root (`sudo ./sdp_i2c_example_usage`), it's
    typically due to wrong permission settings. See the next chapter how to solve this.

### Missing I2C permissions

If your user is missing access to the I2C interface you should first verfiy
the user belongs to the `i2c` group.

```
$ groups
users input some other groups etc
```
If `i2c` is missing in the list add the user and restart the Raspberry Pi.

```
$ sudo adduser your-user i2c
Adding user `your-user' to group `i2c' ...
Adding user your-user to group i2c
Done.
$ sudo reboot
```

If that did not help you can make globally accessible hardware interfaces
with a udev rule. Only do this if everything else failed and you are 
reasoably confident you are the only one having access to your Pi.

Go into the `/etc/udev/rules.d` folder and add a new file named 
`local.rules`.
```
$ cd /etc/udev/rules.d/
$ sudo touch local.rules
```
Then add a single line `ACTION=="add", KERNEL=="i2c-[0-1]*", MODE="0666"` 
to the file with your favorite editor.
```
$ sudo vi local.rules
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.0] - 2021-08-31

Initial release

[0.1.0]: https://github.com/Sensirion/raspberry-pi-i2c-sdp/releases/tag/0.1.0


sdp_i2c.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-34-gb715f31
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SDP_I2C_H
#define SDP_I2C_H

#ifdef __cplusplus
extern "C" {
#endif

#include "sensirion_config.h"

#define SDP8XX_I2C_ADDRESS_0 0x25
#define SDP8XX_I2C_ADDRESS_1 0x26
#define SDP3X_I2C_ADDRESS_0 0x21
#define SDP3X_I2C_ADDRESS_1 0x22
#define SDP3X_I2C_ADDRESS_2 0x23

#define SDP_I2C_ADDRESS SDP8XX_I2C_ADDRESS_0

/**
 * sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging() - This
 * command starts continuous measurements with mass flow temperature
 * compensation and the average till read feature.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t
sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging(void);

/**
 * sdp_start_continuous_measurement_with_mass_flow_t_comp() - This command
 * starts continuous measurements with mass flow temperature compensation.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_start_continuous_measurement_with_mass_flow_t_comp(void);

/**
 * sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging() -
 * This command starts continuous measurements with differential pressure
 * temperature compensation and the average till read feature.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t
sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging(void);

/**
 * sdp_start_continuous_measurement_with_diff_pressure_t_comp() - This command
 * starts continuous measurements with differential pressure temperature
 * compensation.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_start_continuous_measurement_with_diff_pressure_t_comp(void);

/**
 * sdp_stop_continuous_measurement() - This command stops the continuous
 * measurement and puts the sensor in idle mode. It powers off the heater and
 * makes the sensor receptive to another command after 500us. The Stop command
 * is also required when switching between different continuous measurement
 * commands.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_stop_continuous_measurement(void);

/**
 * sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging() - This command
 * triggers a single shot measurement with mass flow temperature compensation.
 *
 * @note During a triggered measurement the sensor measures both differential
 * pressure and temperature. The measurement starts directly after the command
 * has been sent. The command needs to be repeated with every measurement.
 * During the 45ms that the sensor is measuring, no command can be sent to the
 * sensor. After the 45ms the result can be read out and any command can be sent
 * to the sensor.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging(void);

/**
 * sdp_trigger_measurement_with_diff_pressure_t_comp() - This command triggers a
 * single shot measurement with differential pressure temperature compensation.
 *
 * @note During a triggered measurement the sensor measures both differential
 * pressure and temperature. The measurement starts directly after the command
 * has been sent. The command needs to be repeated with every measurement.
 * During the 45ms that the sensor is measuring, no command can be sent to the
 * sensor. After the 45ms the result can be read out and any command can be sent
 * to the sensor.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_trigger_measurement_with_diff_pressure_t_comp(void);

/**
 * sdp_read_measurement_raw() - After a start continuous measurement commands,
 * the measurement results can be read out at most every 0.5ms. After a
 * triggered measurement command, the results can be read out when the sensor is
 * finished with the measurement. This method returns raw sensor values.
 * The temperature and scale factor don’t need to be read out (every time).
 * The read sequence can be aborted by a NACK and a
 * STOP condition. The scale factor is for differential pressure in Pascal.
 *
 * @param differential_pressure The digital calibrated differential pressure
 * signal read from the sensor is a signed integer number (two's complement
 * number). The integer value can be converted to the physical value by dividing
 * it by the scale factor. differential pressure in Pascal = sensor output /
 * scale factor
 *
 * @param temperature The digital calibrated temperature signal read from the
 * sensor is a signed integer number (two's complement number). The integer
 * value can be converted to the physical value by dividing it by a scale factor
 * of 200. temperature in °C = sensor output / 200
 *
 * @param scaling_factor Scaling factor differential pressure
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_read_measurement_raw(int16_t* differential_pressure,
                                 int16_t* temperature, int16_t* scaling_factor);

/**
 * sdp_read_measurement() - After a start continuous measurement commands, the
 * measurement results can be read out at most every 0.5ms. After a triggered
 * measurement command, the results can be read out when the sensor is finished
 * with the measurement. This method returns the measurement values in physical
 * units.
 *
 * @param differential_pressure measurement value for differential pressure in
 * Pascal
 *
 * @param temperature measurement value for temperature in °C
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_read_measurement(float* differential_pressure, float* temperature);

/**
 * sdp_enter_sleep_mode() - In sleep mode the sensor uses the minimum amount of
 * current. The mode can only be entered from idle mode, i.e. when the sensor is
 * not measuring. This mode is particularly useful for battery operated devices.
 * To minimize the current in this mode, the complexity of the sleep mode
 * circuit has been reduced as much as possible, which is mainly reflected by
 * the way the sensor exits the sleep mode. In sleep mode the sensor cannot be
 * soft reset.
 *
 * @note Triggered mode: the sleep command can be sent after the result have
 * been read out and the sensor is in idle mode. Continuous mode: the sleep
 * command can be sent after a stop continuous measurement command has been
 * issued and the sensor is in idle mode.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_enter_sleep_mode(void);

/**
 * sdp_exit_sleep_mode() - Exit sleep mode. See the data sheet for more detailed
 * information
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_exit_sleep_mode(void);

/**
 * sdp_prepare_product_identifier() - Prepare for reading the product identifier
 * and sensor serial number.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_prepare_product_identifier(void);

/**
 * sdp_read_product_identifier() - Read the product identifier and serial number
 * of the sensor.
 *
 * @note Make sure to call 'prepare product identifier' immediately before.
 *
 * @param product_number 32 bit unique product and revision number. The number
 * is listed in the datasheet. Note that the last 8 bits are the revision number
 * and can be subject to change.
 *
 * @param serial_number 64-bit unique serial number
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_read_product_identifier(uint32_t* product_number,
                                    uint8_t* serial_number,
                                    uint8_t serial_number_size);

#ifdef __cplusplus
}
#endif

#endif /* SDP_I2C_H */


sdp_i2c_example_usage.c
/*
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-34-gb715f31
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>  // printf

#include "sdp_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c_hal.h"
#include <inttypes.h>

/*
 * TO USE CONSOLE OUTPUT (PRINTF) YOU MAY NEED TO ADAPT THE INCLUDE ABOVE OR
 * DEFINE IT ACCORDING TO YOUR PLATFORM:
 * #define printf(...)
 */

int main(void) {
    int16_t error = 0;

    sensirion_i2c_hal_init();

    sdp_stop_continuous_measurement();
    // ignore error if no measurement running

    uint32_t product_number;
    const uint8_t serial_number_size = 8;
    uint8_t serial_number_raw[serial_number_size];

    error = sdp_prepare_product_identifier();
    if (error) {
        printf("Error executing sdp_prepare_product_identifier(): %i\n", error);
    }

    error = sdp_read_product_identifier(&product_number, serial_number_raw,
                                        serial_number_size);

    if (error) {
        printf("Error executing sdp_read_product_identifier(): %i\n", error);
    } else {
        printf("Product number: 0x%08x\n", product_number);
        uint64_t serial_number = (uint64_t)serial_number_raw[0] << 56 |
                                 (uint64_t)serial_number_raw[1] << 48 |
                                 (uint64_t)serial_number_raw[2] << 40 |
                                 (uint64_t)serial_number_raw[3] << 32 |
                                 (uint64_t)serial_number_raw[4] << 24 |
                                 (uint64_t)serial_number_raw[5] << 16 |
                                 (uint64_t)serial_number_raw[6] << 8 |
                                 (uint64_t)serial_number_raw[7];
        printf("Serial Number: %" PRIu64 "\n", serial_number);
    }

    // Start Measurement

    error =
        sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging();
    if (error) {
        printf("Error executing "
               "sdp_start_continuous_measurement_with_diff_pressure_t_comp_"
               "and_averaging(): %i\n",
               error);
    }

    for (int i = 0; i < 10000; i++) {
        // Read Measurement every 10ms
        sensirion_i2c_hal_sleep_usec(10000);

        float differential_pressure;
        float temperature;

        error = sdp_read_measurement(&differential_pressure, &temperature);

        if (error) {
            printf("Error executing sdp_read_measurement(): %i\n", error);
        } else {
            printf("Differential pressure: %0.2f Pa\n", differential_pressure);
            printf("Temperature: %0.2f °C\n", temperature);
        }
    }

    error = sdp_stop_continuous_measurement();
    if (error) {
        printf("Error executing sdp_stop_continuous_measurement(): %i\n",
               error);
    }

    return 0;
}


sdp_i2c.c
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-34-gb715f31
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sdp_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"

#define TEMPERATURE_DIVISION_FACTOR 200

static float convert_temperature_raw_to_celsius(int16_t temperature_raw) {
    return (float)temperature_raw / TEMPERATURE_DIVISION_FACTOR;
}

int16_t
sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3603);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_start_continuous_measurement_with_mass_flow_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3608);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t
sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3615);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_start_continuous_measurement_with_diff_pressure_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x361E);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_stop_continuous_measurement(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3FF9);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(500);
    return NO_ERROR;
}

int16_t sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3624);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(45000);
    return NO_ERROR;
}

int16_t sdp_trigger_measurement_with_diff_pressure_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x362F);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(45000);
    return NO_ERROR;
}

int16_t sdp_read_measurement_raw(int16_t* differential_pressure,
                                 int16_t* temperature,
                                 int16_t* scaling_factor) {
    int16_t error;
    uint8_t buffer[9];

    error = sensirion_i2c_read_data_inplace(SDP_I2C_ADDRESS, &buffer[0], 6);
    if (error) {
        return error;
    }
    *differential_pressure = sensirion_common_bytes_to_int16_t(&buffer[0]);
    *temperature = sensirion_common_bytes_to_int16_t(&buffer[2]);
    *scaling_factor = sensirion_common_bytes_to_int16_t(&buffer[4]);
    return NO_ERROR;
}

int16_t sdp_read_measurement(float* differential_pressure, float* temperature) {
    int16_t error;
    int16_t differential_pressure_raw;
    int16_t temperature_raw;
    int16_t scaling_factor;

    error = sdp_read_measurement_raw(&differential_pressure_raw,
                                     &temperature_raw, &scaling_factor);
    if (error) {
        return error;
    }

    *differential_pressure = (float)differential_pressure_raw / scaling_factor;
    *temperature = convert_temperature_raw_to_celsius(temperature_raw);
    return NO_ERROR;
}

int16_t sdp_enter_sleep_mode(void) {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3677);

    return sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
}

int16_t sdp_exit_sleep_mode(void) {
    uint8_t buffer[2];
    uint16_t offset = 0;

    sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    // ignore error, as wakeup is not acknowledged
    sensirion_i2c_hal_sleep_usec(2000);
    return NO_ERROR;
}

int16_t sdp_prepare_product_identifier(void) {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x367C);

    return sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
}

int16_t sdp_read_product_identifier(uint32_t* product_number,
                                    uint8_t* serial_number,
                                    uint8_t serial_number_size) {
    int16_t error;
    uint8_t buffer[18];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xE102);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(1000);

    error = sensirion_i2c_read_data_inplace(SDP_I2C_ADDRESS, &buffer[0], 12);
    if (error) {
        return error;
    }
    *product_number = sensirion_common_bytes_to_uint32_t(&buffer[0]);
    sensirion_common_copy_bytes(&buffer[4], serial_number, serial_number_size);
    return NO_ERROR;
}
